### ISSUE TITLE
npm package for pyodide

### ISSUE THREAD

**Issue by Kreijstal** [pyodide/pyodide#328]
I've seen there has been a lot of progress with this project, any ideas about making an npm package about it, soon?

**Comment by mdboom** [comment]
That's a great idea.  I don't know if I will have the bandwidth soon, but I certainly would be open to pull requests.

**Comment by rth** [comment]
There is a in implementation in https://github.com/iodide-project/pyodide/pull/792 though it would require some work to integrate it.

**Comment by rth** [comment]
So we are progressively moving toward a npm package for pyodide.js thanks to @hoodmane  (https://github.com/pyodide/pyodide/pull/1575, https://github.com/pyodide/pyodide/pull/1578). 

@jtpio @bollwyvl If you have any thoughts on this or design constraints given your work in jupyterlight please let us know.

**Comment by bollwyvl** [comment]
epic work, :clap: @hoodmane :cake: 

aww, thanks for thinking about us. @jtpio may have more thoughts, but here are some off the top of my head, without much filter, so watch your toes:

I don't know if we have a lot of _constraints_ other than "works in browser," which i assume is table stakes.

With webpack 5, there are a lot fewer worries about things, as it's much easier to _copy a file into place_ and have it named correctly and predictably, without webpack even looking at it. See OOM errors, etc.

The biggest thing that would help us, specifically, would be if it shipped with first-party `"types": "lib/index.d.ts"`, so that we could more confidently upgrade inside of an already-complex monstrosity, but that may be out-of-scope for an initial release, as types that aren't tested are probably worse than no types at all. The lightest touch for this would be jsdoc annotations which can [generate the required file](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html).

I don't know if the `.data` files get served correctly by all web hosts as wasm, but I actually haven't seen a lot of problems with it. I don't know how packaging can help, more of a hosting problem, but then `application/wasm` is _still_ provisional, i think...

Sooner rather than later, `jupyterlite` is gonna need (optional, but strongly recommended) ServiceWorker support, and not just because of pyodide wasm loading/caching. This might be something that pyodide could offer (or detect) natively, as an alternate build in the same dist. I haven't dug into it much, but this seems like the biggest win for cutting down both time-to-first interaction and memory footprint in "production" deployments, especially ones with multiple tabs. The are rather complicated to make work in all hosting settings, so maybe best to leave it to downstreams, but maybe have some docs.

It's gross, but I have the strong suspicion we're going to end up shipping python `.tar.gz` and `.whl` and then eventually conda `.tar.bz2` of pyodide, having gone through the jupyterlab webpack 5 federated thing so users don't have to run it locally (or even have node). We might have to inject some cache-busting in the URL paths (usually `[version]` is enough) but managing the maximum length of file names is important for windows users... jupyterlab isn't making things better, as it ships with a minimum prefix of `{sys.prefix}/share/jupyter/labextensions/@full-npm-org/full-npm-name/static/`. Not sure if this has much bearing here, but is the kind of thing that keeps me up at night.

Thanks for asking/listening!

**Comment by rth** [comment]
Thanks a lot for the input @bollwyvl !

> The biggest thing that would help us, specifically, would be if it shipped with first-party "types": "lib/index.d.ts", so that we could more confidently upgrade

I don't have much experience with TypeScript, but it sounds reasonable to move in that direction. That's what https://github.com/pyodide/pyodide/pull/792 did.

> I don't know if the .data files get served correctly by all web hosts as wasm, but I actually haven't seen a lot of problems with it. 

I think, it's not technically necessary, `octet-stream` also works for packages `.data`. We did it mostly so it would be compressed by default by the CDN https://github.com/pyodide/pyodide/issues/671 and I'm not sure if it's still relevant with JsDelivr. 

> ServiceWorker support, and not just because of pyodide wasm loading/caching. This might be something that pyodide could offer (or detect) natively, as an alternate build in the same dist. I haven't dug into it much, but this seems like the biggest win for cutting down both time-to-first interaction and memory footprint in "production" deployments, especially ones with multiple tabs

We should open a separate issue about it.

**Comment by bollwyvl** [comment]
#792 looks heavy! Frankly, it doesn't bother me that `pyodide.js` can't be webpacked or whatever. 

When I mention the "light touch" of the jsdoc idea: it's add _just enough_ structured documentation to the js that the typechecker can hold a function accountable... and also makes great automated docs, which feels nice.

Looks like @hoodmane is way ahead of us (me) on the serviceworkers: https://github.com/pyodide/pyodide/issues/1545#issuecomment-828659003

Anyhow, not sure when I'll be able to throw any weight at the jsdoc ideas, but let us know if there's anything you'd like us to guinea pig... we're definitely _not_ in the "shipping stuff" (other than docs and a site, yet) stage, so it's pretty low risk for us to try things out.

**Comment by hoodmane** [comment]
This jsdoc idea sounds really good to me, I had no idea that was an option.

One big advantage of the jsdoc approach is that we should be able to apply this also to generate types for the public `EM_JS` functions, most importantly for the entire PyProxy API. It will probably be a long time before Emscripten supports embedded typescript, but we are already tricking jsdoc into processing `EM_JS` code as valid Javascript so I bet the type generation would work as-is. Though it might be more sensible to switch to using rollup to bundle the current `EM_JS` javascript files, in which case we'd have more flexibility (in particular, we could escape from the annoying "syntax error on line 10 character 43720" errors and the limitation to only use the old js syntax acceptable to the tools that emscripten uses to transform `EM_JS` code).

**Comment by jtpio** [comment]
> The biggest thing that would help us, specifically, would be if it shipped with first-party `"types": "lib/index.d.ts"`,

Yes, this would already greatly help and improve the developer experience :+1:

**Comment by hoodmane** [comment]
I made a basic attempt at this in #1592.

**Comment by bollwyvl** [comment]
Those are looking really good to start with. I already love seeing the `Int32Array` in there, and the docs are looking very nice in `api.d.ts`.

On their [hints](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html#tips), the offiicial docs suggest [tsd](https://github.com/SamVerschueren/tsd). This approach is also used on, e.g. DefinitelyTyped, and seems to work for them at scale. These files _also_ don't even get transpiled, apparently, which is dark magic indeed.

Another form of testing the types (by inspection) might generating the docs: on [jupyterlite](https://jupyterlite.readthedocs.io/en/latest/api/ts/classes/pyolite_kernel_src.pyolitekernel-1.html) we're using `typedoc` to generate markdown, and then include that in a `sphinx` (with `myst-nb`) build. In turn, running a link check on the output of _that_ is quite effective to find if some narrative reference to the functions/classes (or higher types) is out-of-date.

**Comment by bollwyvl** [comment]
Oh on the service worker front, did some reading:

> _In order to facilitate local development, `localhost` is considered a
secure origin by browsers as well._

>
https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers#setting_up_to_play_with_service_workers

So this makes it less crazy to be able to debug while building. I would
still imagine that it would need to have a fall-back, as http hosts still
do exist in e.g. intranet-over-vpn, even if it's not best practice.

**Comment by hoodmane** [comment]
My experience with service workers was that Firefox didn't consider localhost to be a secure origin and that Chrome wasn't implementing the secure origin rules at all.

**Comment by hoodmane** [comment]
tsd seems quite helpful, but like many node packages it has frustratingly specific opinions about directory structure and file names.

**Comment by hoodmane** [comment]
Okay so I got things working pretty well. Here is my test file:
https://github.com/hoodmane/pyodide/blob/types-2/src/js/index.test-d.ts

It'll probably take a few PRs to get the types added in (that branch has a pretty large diff though two thirds of the diff is the entire file pyproxy.js changing indentation level).

Calling Python methods will not be all that ergonomic. For example, suppose `p` is a proxy and I want to call `p.pymethod()`. The options are a type assertion:
```ts
(p.pymethod as PyProxyCallable)();
```
or a pair of type guards:
```ts
let f = p.pymethod;
assert(pyodide.isPyProxy(f) && f.isCallable());
f();
```
I suppose this is what you expect though.

**Comment by bollwyvl** [comment]
Yeah, once it gets to a certain point, if something upstream (in this case all of python) is underspecified, there's not much one can do! 

I think if a downstream wanted to document their findings of a particular library, they'd do cast `p` to something they wanted to work with, e.g. 

```ts
export interface IMyStringInterface {
  strip(): IMyStringInterface;
  lower(): IMyStringInterface;
}
const pTheString = p as IMyStringInterface; // but this can always go wrong
pTheString.strip().lower();
```

Presumably it would be feasible to automatically generate such types for at least stdlib from typeshed, but now we're getting _way_ off the beaten path... perhaps in some bright future, there will be a cross-languge specification for an algebra of types, e.g. [WebAssembly Interface Types](https://hacks.mozilla.org/2019/08/webassembly-interface-types) but with higher-order constructs, but for now we do what we can!

**Comment by bollwyvl** [comment]
Another thing that [came up](https://github.com/jtpio/jupyterlite/discussions/91#discussioncomment-754385): I don't know magnitude of the work/rebuilds required, but for jupyterlite purposes it would be _really nice_ if there was either _a_ build or _the_ build which used the [IDBFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs) backend. Naively, it seems like this would "only" require `-lidbfs.js` next to [`FORCE_FILESYSTEM=1`](https://github.com/pyodide/pyodide/blob/main/Makefile#L50)...

**Comment by hoodmane** [comment]
The most obvious concern I'd have about switching to IDBFS is that unlike the current FS we're using it is persistent. The Pyodide startup logic initializes the filesystem with a bunch of stuff. But that is all generated by emscripten which presumably knows whether it's been asked to use a persistent or a transient filesystem. When using a persistent filesystem, perhaps it checks on startup if the file system has been initialized already and in that case skips initialization?

If you try it, please let us know how it works out!

Of course the holy grail for this would be a [File System Access API](https://web.dev/file-system-access/)-backed FS, because then not only would the install persist but files stored into the Pyodide filesystem would be directly mapped to a location on the user's hard drive for easy interoperability with other programs. Too bad firefox doesn't seem too interested in supporting it... (https://github.com/mozilla/standards-positions/issues/154)

**Comment by bollwyvl** [comment]
Welp, adding `-lidbfs.js` didn't break the build, and then tried turning all of them on, with a similarly non-spectacular continuing-to-function console page in the docs.

As we don't care about `/tmp` and `/dev`, I think in pyodide, our worker initialization, or... _somewhere_ we'd do something like:

```ts
FS.mount(IDBFS, {}, "/home/jovyan");
```

What I readily know how to do is:
- get a handle on `FS` and `IDBFS`
- specify a set of mount points to pyodide at initialization/runtime
  - our filesystem will exist prior (and after) the worker
- write/read files to said filesystem that are formatted correctly, but there's likely some handy library out there 
- hook up listeners on either/both the pyodide or jupyterlite side of the house to report when the files change
  - we already have some slow-polling on our side of the house

Seems promising though, and persisting _just_ our data is far preferable to the whole fs anyway!

**Comment by bollwyvl** [comment]
Well, it almost certainly isn't the right spot, but adding this awful hack to `main.c`...
```c
int
main(int argc, char** argv)
{
	EM_ASM({
		FS.mkdir('/home/jo');
		FS.mount(IDBFS, {}, '/home/jo');
                FS.syncfs(true, function (err) {
                  err && console.error('error syncing FROM IndexedDB', err);
                });
                setInterval(function() {
                   FS.syncfs(false, function (err) {
                      err && console.error('error syncing TO IndexedDB', err);
                   });
                }, 5000);
	});
```

...does yield persistent files, across tabs, etc:

![Screenshot from 2021-05-19 23-34-01](https://user-images.githubusercontent.com/45380/118915295-b6044500-b8fa-11eb-98ff-48c6ccca5694.png)

So aside from figuring out when/how to apply `FS.syncfs` (and in what direction) on the pyodide side, and then having some other channel for signaling/locking with our existing client-facing files, this could probably be made to work. Further, `micropip` could probably make use of this for a wheelhouse, user packages, etc.

The question definitely still remains whether this could be dynamically configured... and how we might coordinate this with other esmcripten-based language runtimes.

**Comment by hoodmane** [comment]
Adding this somewhere in `src/js` would probably be more reasonable -- `EM_ASM` has some weird behaviors and requires rerunning the linker if you change things.

You can access `FS` as `Module.FS` inside `pyodide.js`, though only after
https://github.com/pyodide/pyodide/blob/main/src/js/pyodide.js#L197
You will presumably have to replace `IDBFS` with `Module.IDBFS` too.
